packetbeat/scripts/mage/build.go

```diff
@@ -17,11 +17,50 @@
 
 package mage
 
-import devtools "github.com/elastic/beats/v7/dev-tools/mage"
+import (
+	"os"
+	"strings"
+
+	devtools "github.com/elastic/beats/v7/dev-tools/mage"
+)
+
+// NpcapVersion specifies the version of the OEM Npcap installer to bundle with
+// the packetbeat executable. It is used to specify which npcap builder crossbuild
+// image to use.
+const NpcapVersion = "1.60"
+
+// BeatsCIHostPrefix is the prefix of hostnames used during beats builds.
+const BeatsCIHostPrefix = "beats-ci-immutable"
 
 // CrossBuild cross-builds the beat for all target platforms.
+//
+// On Windows platforms, if CrossBuild is invoked with the environment variables
+// NPCAP_LOCAL set to "true" or the builder hostname has the BeatsCIHostPrefix,
+// a private cross-build image is selected that provides the OEM Npcap installer
+// for the build. This behaviour requires access to the private image.
 func CrossBuild() error {
-	// Run all builds serially to try to address failures that might be caused
-	// by concurrent builds. See https://github.com/elastic/beats/issues/24304.
-	return devtools.CrossBuild(devtools.Serially())
+	return devtools.CrossBuild(
+		// Run all builds serially to try to address failures that might be caused
+		// by concurrent builds. See https://github.com/elastic/beats/issues/24304.
+		devtools.Serially(),
+
+		devtools.ImageSelector(func(platform string) (string, error) {
+			image, err := devtools.CrossBuildImage(platform)
+			if err != nil {
+				return "", err
+			}
+			hostname, err := os.Hostname()
+			if err != nil {
+				return "", err
+			}
+			if !strings.HasPrefix(hostname, BeatsCIHostPrefix) && os.ExpandEnv("NPCAP_LOCAL") != "true" {
+				return image, nil
+			}
```

Comment:
Are you happy with this as an approach to checking whether we are on CI or a dev's machine?

Reply:
`CI` env variable is set by default in almost all the CI vendors as true. So I'd say to use this env variable instead. See https://issues.jenkins.io/browse/JENKINS-36707 as a reference for the Jenkins specific case

Reply:
I tried that in a previous experiment since you mentioned that in the meeting, but it did not work; it never got to the point of the mage build script where it could set the npcap image target. This is the reason for the rigmarole of examining the env vars in the experiments above, for example [here](https://github.com/elastic/beats/commit/9dd6bbfbc45babaf63b70de9400ffbbf44d3cfbf).

Reply:
I'm not familiar with the `os.ExpandEnv` but does it run within the host or within a docker container? If the latter, how does the `mage package` pass the environment variable context from the host to the container? In case, it's a matter of env propagation.

https://beats-ci.elastic.co/pipeline-syntax/globals#env contains the list of available env variables

![image](https://user-images.githubusercontent.com/2871786/150978413-c39b1122-a4ed-4b90-8947-7ed018f5bed3.png)

Reply:
`CrossBuild` should be running on the CI host. It executes prior to the docker container in order to select the appropriate image. I think we should depend on CI=true and avoid the hostname.

Reply:
The section of code that looks at the env to determine the image to pull is run in the host and so should see $CI if it's properly exported.

I've repeated that [experiment](https://github.com/elastic/beats/pull/29112/commits/010de73f1a9b1d674af1f7ccfb6b39bf35486330) and this is the result in x-pack/packetbeat-packaging-linux.

```
[2022-01-25T14:31:17.237Z] Digest: sha256:f58be865603aa58d5191c026134d1d462d2d4525f8138482d09eb41ded760642
[2022-01-25T14:31:17.237Z] Status: Downloaded newer image for docker.elastic.co/beats-dev/golang-crossbuild:1.17.5-darwin-debian10
[2022-01-25T14:31:19.150Z] >> Building using: cmd='build/mage-linux-amd64 golangCrossBuild', env=[CC=o64-clang, CXX=o64-clang++, GOARCH=amd64, GOARM=, GOOS=darwin, PLATFORM_ID=darwin-amd64]
[2022-01-25T14:32:26.918Z] >> golangCrossBuild: Building for linux/386
[2022-01-25T14:32:26.919Z] >> Building using: cmd='build/mage-linux-amd64 golangCrossBuild', env=[CC=gcc, CXX=g++, GOARCH=386, GOARM=, GOOS=linux, PLATFORM_ID=linux-386]
[2022-01-25T14:33:34.693Z] >> golangCrossBuild: Building for linux/amd64
[2022-01-25T14:33:34.694Z] >> Building using: cmd='build/mage-linux-amd64 golangCrossBuild', env=[CC=gcc, CXX=g++, GOARCH=amd64, GOARM=, GOOS=linux, PLATFORM_ID=linux-amd64]
[2022-01-25T14:34:42.476Z] >> golangCrossBuild: Building for linux/arm64
[2022-01-25T14:34:42.476Z] >> Building using: cmd='build/mage-linux-amd64 golangCrossBuild', env=[CC=aarch64-linux-gnu-gcc, CXX=aarch64-linux-gnu-g++, GOARCH=arm64, GOARM=, GOOS=linux, PLATFORM_ID=linux-arm64]
[2022-01-25T14:35:50.259Z] >> golangCrossBuild: Building for windows/386
[2022-01-25T14:35:50.259Z] >> Building using: cmd='build/mage-linux-amd64 golangCrossBuild', env=[CC=i686-w64-mingw32-gcc, CXX=i686-w64-mingw32-g++, GOARCH=386, GOARM=, GOOS=windows, PLATFORM_ID=windows-386]
[2022-01-25T14:35:50.260Z] Error: failed to copy Npcap installer into source tree: can't copy /installer/npcap-1.60-oem.exe: open /installer/npcap-1.60-oem.exe: no such file or directory
[2022-01-25T14:35:50.260Z] Error: failed building for windows/386: exit status 1
[2022-01-25T14:35:50.260Z] failed building for windows/386: exit status 1
[2022-01-25T14:35:50.260Z] package ran for 5m24.576511921s
[2022-01-25T14:35:50.260Z] Error: failed cross-building target=golangCrossBuild for platform=windows/386: running "docker run --env EXEC_UID=1164 --env EXEC_GID=1165 -v /var/lib/jenkins/workspace/PR-29112-64-f28b08c4-1499-4d14-a721-0404ce4b1283/pkg/mod:/go/pkg/mod:ro --rm --env GOFLAGS=-mod=readonly --env MAGEFILE_VERBOSE= --env MAGEFILE_TIMEOUT= --env SNAPSHOT=true --env DEV=false -v /var/lib/jenkins/workspace/PR-29112-64-f28b08c4-1499-4d14-a721-0404ce4b1283/src/github.com/elastic/beats:/go/src/github.com/elastic/beats -w /go/src/github.com/elastic/beats/x-pack/packetbeat docker.elastic.co/beats-dev/golang-crossbuild:1.17.5-main --build-cmd build/mage-linux-amd64 golangCrossBuild --platforms windows/386" failed with exit code 1
script returned exit code 1
```

You can see here that the standard image, docker.elastic.co/beats-dev/golang-crossbuild:1.17.5-main, is being pulled resulting in the failure due to the absence of the installer asset. This happens (from previous experiments not shown here) because if the CI var is being set by the builder, it is not being exported to the mage script executable and so is not visible. Unfortunately I can't link to logs showing this.

Reply:
Question, what's the docker image that has been used? when 

points to 


I just committed https://github.com/elastic/beats/pull/29112/commits/b2ee5126f196fd464c7dc4d2135827994f1e4150
so there should be a few references with the `CI` if it's set by Jenkins.

1) The one before the git checkout.
2) The one in the `dumpVariables` that happens in each Stage.
3) The one in the packaging stage.

for 1) is there:

![image](https://user-images.githubusercontent.com/2871786/151058167-649e95ab-e607-408e-b99c-5203a55ae68a.png)

for 2) is also there:

![image](https://user-images.githubusercontent.com/2871786/151059767-09827624-5067-4bbe-8993-5dd815caa1cf.png)

for 3) we just need to wait a bit longer

I'd say at the moment the CI=true is correctly set and available in each worker. We can wait for `3)` to verify this.

OTOH, I just ran the below commands with a similar outcome:

```bash
cd x-pack/packetbeat
CI=true mage package
```

```
>> Building using: cmd='build/mage-linux-amd64 golangCrossBuild', env=[CC=i686-w64-mingw32-gcc, CXX=i686-w64-mingw32-g++, GOARCH=386, GOARM=, GOOS=windows, PLATFORM_ID=windows-386]
Error: failed to copy Npcap installer into source tree: can't copy /installer/npcap-1.60-oem.exe: open /installer/npcap-1.60-oem.exe: no such file or directory
Error: failed building for windows/386: exit status 1
failed building for windows/386: exit status 1
package ran for 9m45.589113394s
```

Though, I see a weird trace:

```
Unable to find image 'docker.elastic.co/beats-dev/golang-crossbuild:1.17.2-main' locally
1.17.2-main: Pulling from beats-dev/golang-crossbuild
2f0ef4316716: Already exists
2039af7edb8e: Already exists
030257de1373: Already exists
b61458b27a37: Already exists
f65dcd0bdecc: Already exists
4ff18e112eed: Already exists
6915df32d574: Already exists
855a22c41e25: Already exists
d89370b94b16: Already exists
6e40c4c56e97: Pulling fs layer
f5bea9cc5a1b: Pulling fs layer
9f151d97cbbb: Pulling fs layer
7ddfbeb14bb5: Pulling fs layer
...
f1ff61e1a288: Pull complete
Digest: sha256:0c59d3fe423f5db78c94a07372cccf2b9c1a8f7d1cf19fcfa34efc102ef2c362
Status: Downloaded newer image for docker.elastic.co/beats-dev/golang-crossbuild:1.17.2-main
```

I don't know where `1.17.2` is coming from as I'm using the changes in this PR:

```
git --no-pager log -2
commit b2ee5126f196fd464c7dc4d2135827994f1e4150 (HEAD -> npcap/installer)
Author: Victor Martinez <VictorMartinezRubio@gmail.com>
Date:   Tue Jan 25 20:48:31 2022 +0000

    test: print the CI env variable

commit 010de73f1a9b1d674af1f7ccfb6b39bf35486330
Author: Dan Kortschak <dan.kortschak@elastic.co>
Date:   Tue Jan 25 22:27:51 2022 +1030

    experiment DO NOT MERGE
```

Does it work in your end?

Reply:
It may be worth dropping this in
```diff
diff --git a/dev-tools/mage/settings.go b/dev-tools/mage/settings.go
index b721b730c6..fbcc916fbe 100644
--- a/dev-tools/mage/settings.go
+++ b/dev-tools/mage/settings.go
@@ -209,6 +209,7 @@ BeatUser         = {{.BeatUser}}
 VersionQualifier = {{.Qualifier}}
 PLATFORMS        = {{.PLATFORMS}}
 PACKAGES         = {{.PACKAGES}}
+CI               = {{.CI}}
 
 ## Functions
 
```

Reply:
That 1.17.2 is coming in because of the age of the PR, I'll rebase onto master when your build has finished, and add the CI dump above.

Reply:
Test failed prior to packaging, the log is [here](https://beats-ci.elastic.co/job/Beats/job/beats/job/PR-29112/65/consoleText).

Reply:
This confirms that CI is indeed set correctly, and the mage script does have access to it.
```
[2022-01-26T00:37:20.937Z] ### MAGE DUMP ###
[2022-01-26T00:37:20.959Z] [Pipeline] isUnix
[2022-01-26T00:37:20.962Z] [Pipeline] sh
[2022-01-26T00:37:21.249Z] + mage dumpVariables
[2022-01-26T00:37:33.543Z] ## Variables
[2022-01-26T00:37:33.543Z] 
[2022-01-26T00:37:33.543Z] GOOS             = linux
[2022-01-26T00:37:33.543Z] GOARCH           = amd64
[2022-01-26T00:37:33.543Z] GOARM            = 
[2022-01-26T00:37:33.543Z] Platform         = linux/amd64
[2022-01-26T00:37:33.543Z] BinaryExt        = 
[2022-01-26T00:37:33.543Z] XPackDir         = ../x-pack
[2022-01-26T00:37:33.543Z] BeatName         = beats
[2022-01-26T00:37:33.543Z] BeatServiceName  = beats
[2022-01-26T00:37:33.543Z] BeatIndexPrefix  = beats
[2022-01-26T00:37:33.543Z] BeatDescription  = 
[2022-01-26T00:37:33.543Z] BeatVendor       = Elastic
[2022-01-26T00:37:33.543Z] BeatLicense      = ASL 2.0
[2022-01-26T00:37:33.543Z] BeatURL          = https://www.elastic.co/beats/beats
[2022-01-26T00:37:33.543Z] BeatUser         = root
[2022-01-26T00:37:33.543Z] VersionQualifier = 
[2022-01-26T00:37:33.543Z] PLATFORMS        = 
[2022-01-26T00:37:33.543Z] PACKAGES         = 
[2022-01-26T00:37:33.543Z] CI               = true
```
However, it is still not setting the correct image; choosing docker.elastic.co/beats-dev/golang-crossbuild:1.17.5-main.
```
[2022-01-26T02:26:22.841Z] >> Building using: cmd='build/mage-linux-amd64 golangCrossBuild', env=[CC=i686-w64-mingw32-gcc, CXX=i686-w64-mingw32-g++, GOARCH=386, GOARM=, GOOS=windows, PLATFORM_ID=windows-386]
[2022-01-26T02:26:22.841Z] Error: failed to copy Npcap installer into source tree: can't copy /installer/npcap-1.60-oem.exe: open /installer/npcap-1.60-oem.exe: no such file or directory
[2022-01-26T02:26:22.841Z] Error: failed building for windows/386: exit status 1
[2022-01-26T02:26:22.841Z] failed building for windows/386: exit status 1
[2022-01-26T02:26:22.841Z] package ran for 5m16.10085633s
[2022-01-26T02:26:22.841Z] Error: failed cross-building target=golangCrossBuild for platform=windows/386: running "docker run --env EXEC_UID=1164 --env EXEC_GID=1165 -v /var/lib/jenkins/workspace/PR-29112-67-130a0c5b-dc0a-4078-8257-0a113928a8ab/pkg/mod:/go/pkg/mod:ro --rm --env GOFLAGS=-mod=readonly --env MAGEFILE_VERBOSE= --env MAGEFILE_TIMEOUT= --env SNAPSHOT=true --env DEV=false -v /var/lib/jenkins/workspace/PR-29112-67-130a0c5b-dc0a-4078-8257-0a113928a8ab/src/github.com/elastic/beats:/go/src/github.com/elastic/beats -w /go/src/github.com/elastic/beats/x-pack/packetbeat docker.elastic.co/beats-dev/golang-crossbuild:1.17.5-main --build-cmd build/mage-linux-amd64 golangCrossBuild --platforms windows/386" failed with exit code 1
```
So, I look back to find the error I have made. I think I have it.

Reply:
Yes. That is fixed. Trivially incorrect.

@v1v It looks like its failing to copy an artifact though.
```
[2022-01-26T06:02:52.029Z] + gsutil -m -q cp -a public-read packetbeat/build/distributions/**/* gs://beats-ci-artifacts/pull-requests/pr-29112/packetbeat
[2022-01-26T06:02:53.413Z] CommandException: No URLs matched: packetbeat/build/distributions/**/*
[2022-01-26T06:02:53.413Z] CommandException: 1 file/object could not be transferred.
```

Reply:
> @v1v It looks like its failing to copy an artifact though.

As far as I see from the logs, there. are no binaries in the expected folder, does packebeat generate the files in the `build/distributions` folder?

Reply:
The same [failure](https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-30009/runs/4/steps/2390/log/?start=0) is affecting unrelated changes.
```
[2022-01-26T19:19:36.172Z] + gsutil -m -q cp -a public-read auditbeat/build/distributions/**/* gs://beats-ci-artifacts/pull-requests/pr-30009/auditbeat
[2022-01-26T19:19:37.564Z] CommandException: No URLs matched: auditbeat/build/distributions/**/*
[2022-01-26T19:19:37.564Z] CommandException: 1 file/object could not be transferred.
script returned exit code 1
```

Reply:
There are breaking changes in the target master branch related to the packaging :/

Reply:
Thanks. This is fixed by 30007?